src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[2];
+	int param_format[2];
+	int param_length[2];
+
+	std::string old_data = loadBlock(pos);
+
+	// This is of interest. I have noticed there are a large number of
+	// saveBlock calls where the new and old data are exactly the same.
+	if (data.compare(old_data) == 0) {
+		if (old_data.size() > 0) {
+			infostream << "saveBlock called for block already saved, pos = "
+					<< PP(pos) << ", " << data.size() << " bytes"
+					<< std::endl;
+		}
+		return true;
+	}
+
+	if (old_data.size() > 0) {
+		return updateBlock(pos, data);
+	}
+
+	deleteBlock(pos);
+
+	param_values[0] = posTochar(pos);
+	param_format[0] = 0;
+	param_length[0] = strlen(param_values[0]);
+
+	param_values[1] = data.c_str();
+	param_format[1] = 1;
+	param_length[1] = data.size();
+
+	m_res = PQexecPrepared(m_conn, "write_block", 2, (const char *const*)param_values,
+			param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		PQclear(m_res);
+		return true;
+	default:
+		errorstream << "PostgreSQL Error Message:" << PQresultErrorMessage(m_res) << std::endl;
+		PQclear(m_res);
+		return false;
+		break;
+	}
+}
+
+bool Database_PostgreSQL::updateBlock(const v3s16 &pos, const std::string &data)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[2];
+	int param_format[2];
+	int param_length[2];
+
+	param_values[0] = data.c_str();
+	param_format[0] = 1;
+	param_length[0] = data.size();
+
+	param_values[1] = posTochar(pos);
+	param_format[1] = 0;
+	param_length[1] = strlen(param_values[0]);
+
+	m_res = PQexecPrepared(m_conn, "update_block", 2, (const char *const*)param_values,
+			param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		PQclear(m_res);
+		return true;
+	default:
+		errorstream << "PostgreSQL Error Message:" << PQresultErrorMessage(m_res) << std::endl;
+		PQclear(m_res);
+		return false;
+		break;
+	}
+}
+
+std::string Database_PostgreSQL::loadBlock(const v3s16 &pos)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[1];
+	int param_format[1];
+	int param_length[1];
+
+	std::string s   = "";
+
+	param_values[0] = posTochar(pos);
+	param_format[0] = 0;
+	param_length[0] = strlen(param_values[0]);
+
+	m_res = PQexecPrepared(m_conn, "read_block", 1, (const char *const*)param_values, param_length, param_format, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		break;
+	case PGRES_TUPLES_OK:
+		if (PQntuples(m_res))
+			s = std::string(PQgetvalue(m_res, 0,0), PQgetlength(m_res, 0, 0));
+		break;
+	default:
+		errorstream << "PostgreSQL Error:" << PQresultErrorMessage(m_res) << std::endl;
+		break;
+	}
+
+	PQclear(m_res);
+	return s;
+}
+
+const char * Database_PostgreSQL::posTochar(const v3s16 &pos)
+{
+	char buffer[33];
+	snprintf(buffer, 32, "%ld", getBlockAsInteger(pos));
```

Comment:
This is the last sticking point, and I would like to give it the best shot. This seems to give me very good performance also. Please let me know if you seen anything wrong.
        irr::s64 posInt = getBlockAsInteger(pos);

        const char* param_values[]    = {data.c_str(), (const char*)&posInt};
        const int param_format[]      = {1, 1};
        const int param_length[]     = {(int)data.size(), sizeof(irr::s64)};

I also have to test using actual X, Y, and Z, columns in the database, rather than a single pos column. I'll keep plugging away. At the moment chunks are loading fast enough to comfortably around with fast move on.